compiler/rustc_infer/src/infer/combine.rs

```diff
@@ -546,8 +546,7 @@ impl<'tcx> TypeRelation<'tcx> for Generalizer<'_, 'tcx> {
     }
 
     fn mark_ambiguous(&mut self) {
-        // The generalizer always compares types against themselves,
-        // and thus doesn't really take part in coherence.
+        self.infcx.tcx.sess.delay_span_bug(self.cause.span, "opaque types are handled in `tys`");
```

Comment:
because it's reachable in the presence of other errors... whoops the message is wrong now ^^

Reply:
how '^^ shouldn't the generalizer manually match on opaque types in `fn tys` and relat their substs?

Reply:
no I mean, I didn't add that ^^ because that is only reachable if we already error due to the same opaque type being actually compared against itself elsewhere

Reply:
I don't think that's true?

```rust
#![feature(type_alias_impl_trait)]
trait Trait {}
type Opaque<T> = impl Sized;
fn foo<T>() -> Opaque<T> {
    ()
}

impl<T> Trait for (T, u32) {}
impl<U> Trait for (Opaque<U>, i32) {} 
```
this shouldn't error afaict?

Reply:
how does it not cause us to generalize opaque types?

Reply:
we do generalize during coherence, do we?

If so, could we wrap the opaque type, e.g. `(Vec<Opaque<U>>, i32)` or whatever?

Reply:
ah, because of fast reject probably, change it to the following maybe?

```rust
impl<T, V> Trait for (T, V, u32) {}
impl<U, V> Trait for (Opaque<U>, i32, V) {} 
```

Reply:
That example should (and does) error, since these two impls overlap.

I'm not sure how to build an example that should pass but doesn't already get rejected by types_may_unify

Reply:
woops: 
```rust
impl<T, V> Trait for (T, V, V, u32) {}
impl<U, V> Trait for (Opaque<U>, V, i32, V) {} 
```
now `V` has to be the same for both impls